PDSim/core/core.py

```diff
@@ -1131,39 +1131,82 @@ def OBJECTIVE_CYCLE(self, Td_Tlumps0, X, epsilon = 0.003, cycle_integrator = 'RK
                 
             self.solvers.lump_eb_history.append([self.Tlumps, self.lumps_resid])
             
-            if len(self.solvers.lump_eb_history) == 1:
-                
-                T, EB = self.solvers.lump_eb_history[-1]
-                
-                #  T and EB are one-element lists, get floats
-                _T, _EB = T[0], EB[0]
-            
-                # Use the thermal mass to make the step
-                # Here is the logic:
-                # Instantaneous energy balance given by
-                # dU/dt = m*c*(dT/dt) = sum(Qdot)
-                # and if dt = one cycle period (seconds/rev) DeltaT = 2*pi/omega
-                # DELTAT = sum(Qdot)*Deltat/(m*c)
-                thermal_capacitance = 0.49*0.001 # [kJ/K]
-                Deltat = (2*np.pi)/self.omega # [s]
-            
-                #  Update the lump temperatures
-                self.Tlumps = [_T + _EB*Deltat/thermal_capacitance]
-            
+            if len(self.Tlumps) > 1:
+                print("Running multi-lump analysis")
+
+                if self.OEB_solver == 'MDNR':
+                    # Use Multi Dim. Newton Raphson step for multi-lump temperatures
+                    from numpy.linalg import inv
+                    w = 1.0                
+                    dx = 0.5
+                    ytol=1e-3
+                    x = np.array(self.Tlumps,dtype=np.float)
+                    J = np.zeros((len(x),len(x)))
+                    error = 999
+                    # If a float is passed in for dx, convert to a numpy-like list the same shape
+                    # as x
+                    if isinstance(dx,int) or isinstance(dx,float):
+                        dx=dx*np.ones_like(x)
+                                
+                    r0 = np.array(self.lumps_resid)*1000 
+
+                    # Build the Jacobian matrix by columns
+                    for jj in range(len(self.Tlumps)):
+                        delta = np.zeros_like(x)
+                        delta[jj] = dx[jj]
+                        self.Tlumps = self.Tlumps + delta
+                        ri = self.callbacks.lumps_energy_balance_callback()
+                        #print('ri:',ri)
+                        ri = np.array(ri)
+                        J[:,jj] = (ri-r0)/delta[jj]
+
+                    v = np.dot(-inv(J),r0)
```

Comment:
In general it is better to solve J*v =-r0 for v rather than to right-multiply -r0 by inv(J)

Reply:
shall we also change **misc/solvers.py** line 32?

Reply:
Would be best, but optional.

Reply:
I can change it to:  **v = -inv(J)*r0** in both and push the changes within this PR

Reply:
what if I use:   v= np.linalg.solve(J,-r0) ?